pkg/completion/completion.go

```diff
@@ -0,0 +1,136 @@
+// asynchronous computation is completed.
+type Completion interface {
+	// Context returns the context of the asynchronous computation.
+	// If the context is cancelled, e.g. if it times out, the computation must
+	// be cancelled.
+	Context() context.Context
+
+	// Completed notifies of the completion of the asynchronous computation.
+	// Idempotent. Has no effect if the Context is cancelled.
+	Completed()
+}
+
+// WaitGroup waits for a collection of Completions to complete.
+type WaitGroup interface {
+	// Context returns the context of all the Completions in the wait group.
+	Context() context.Context
+
+	// AddCompletion adds a new completion into the wait group and returns it.
+	AddCompletion() Completion
+
+	// Wait blocks until all completions added by calling AddCompletion are
+	// completed, or the context is canceled, whichever happens first.
+	// Returns the context's error if it is cancelled, nil otherwise.
+	Wait() error
+}
+
+// waitGroup waits for a collection of Completions to complete.
+type waitGroup struct {
+	// ctx is the context of all the Completions in the wait group.
+	ctx context.Context
+
+	// counterLocker locks all calls to AddCompletion and Wait, which must not
+	// be called concurrently.
+	counterLocker lock.Mutex
+
+	// pendingCompletions is the list of Completions returned by
+	// AddCompletion() which have not yet been completed.
+	pendingCompletions []*waitGroupCompletion
+}
+
+// NewWaitGroup returns a new WaitGroup using the given context.
+func NewWaitGroup(ctx context.Context) WaitGroup {
+	return &waitGroup{ctx: ctx}
+}
+
+func (wg *waitGroup) Context() context.Context {
+	return wg.ctx
+}
+
+func (wg *waitGroup) AddCompletion() Completion {
+	wg.counterLocker.Lock()
+	defer wg.counterLocker.Unlock()
+	c := &waitGroupCompletion{
+		ctx:  wg.ctx,
+		done: make(chan struct{}),
+	}
+	wg.pendingCompletions = append(wg.pendingCompletions, c)
+	return c
+}
+
+func (wg *waitGroup) Wait() error {
+	wg.counterLocker.Lock()
+	defer wg.counterLocker.Unlock()
+
+Loop:
+	for i, comp := range wg.pendingCompletions {
+		select {
+		case <-comp.done:
+			continue Loop
+		case <-wg.ctx.Done():
+			// Complete the remaining completions to make sure their done
+			// channels are closed.
+			for _, comp := range wg.pendingCompletions[i:] {
+				comp.Completed()
+			}
+			break Loop
+		}
+	}
+	wg.pendingCompletions = nil
+	return wg.ctx.Err()
+}
+
+// waitGroupCompletion is an implementation of Completion that decrements a
+// WaitGroup's counter.
+type waitGroupCompletion struct {
+	// ctx is the context of the wait group.
+	ctx context.Context
+
+	// locker locks accesses to the completion.
+	locker lock.Mutex
+
+	// completed indicates whether Complete has already been called.
+	completed bool
+
+	// done is the channel to be closed when Complete is called the first time.
+	done chan struct{}
+}
+
+func (c *waitGroupCompletion) Context() context.Context {
+	return c.ctx
+}
+
+func (c *waitGroupCompletion) Completed() {
```

Comment:
Oh, if you wanted to avoid the lock here, and the completed variable, we could do this:
```golang
	select {
	case <-c.done:
	default:
		close(c.done)
	}
```

Reply:
Beautiful! Done.